### PR TITLE
Add multi link support

### PR DESCRIPTION
**Pull request by maxime-rainville**
This PRs add support for multi link field via an HasMany relation to Links.

It also adds some extra features like:
- Link Types can now have icons
- Link titles can default to picking up some bit of metadata on their target (e.g: Linking to a CMS page can default to the page title)

# Parent issue
- #44

### CODE REVIEW

File: client/src/components/LinkBox/LinkBox.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import classnames from 'classnames';
+
+const LinkBox = ({ className, children }) => (
```
Comment (maxime-rainville) [inline client/src/components/LinkBox/LinkBox.js:5]: Add some comment to describe what this does.

File: client/src/components/LinkField/LinkField.js [hunk 2]
```diff
@@ -7,7 +7,7 @@ import { inject, injectGraphql, loadComponent } from 'lib/Injector';
 import fieldHolder from 'components/FieldHolder/FieldHolder';
 import PropTypes from 'prop-types';
 
-const LinkField = ({id, loading, Loading, data, LinkPicker, onChange, types, linkDescription, ...props}) => {
+const LinkField = ({id, loading, Loading, data, LinkPicker, onChange, types, linkDescriptions, ...props}) => {
```
Comment (maxime-rainville) [inline client/src/components/LinkField/LinkField.js:10]: `linkDescriptions` is probably not the best description for this anymore.

File: src/Form/MultiLinkField.php [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                $value = json_encode(
+                    array_map(function (Link $link) {
+                        // dd($link);
+                        return $link->jsonSerialize();
+                    }, $list->toArray())
+                );
+            }
+        }
+
+        // dd($value);
```
Comment (maxime-rainville) [inline src/Form/MultiLinkField.php:65]:
```suggestion
```

File: src/Models/EmailLink.php [hunk 4]
```diff
@@ -11,15 +11,20 @@
  */
 class EmailLink extends Link
 {
-
     private static $db = [
         'Email' => 'Varchar(255)'
     ];
 
+    private static $icon = 'block-email';
```
Comment (maxime-rainville) [inline src/Models/EmailLink.php:18]: Need to find a better icon than this one

File: src/Type/Type.php [hunk 5]
```diff
@@ -20,15 +20,20 @@ public function defineLinkTypeRequirements();
...
+    /**
+     * Font icon to attach to this link type
+     */
+    public function LinkTypeIcon(): string;
+
     /**
      * What should be the link description be given this data.
      */
-    public function generateLinkDescription(array $data): string;
+    public function generateLinkDescription(array $data): array;
```
Comment (maxime-rainville) [inline src/Type/Type.php:31]: Might need to rename this.

**Review by mfendeksilverstripe — COMMENTED** [review]
Tested this on vanilla install and it's working well. Really happy to see this addition 👍 

Screen Shot 2022-10-28 at 11 04 02 AM

I'll leave this PR open as it looks like the implementation is not yet finished.

**Comment by edwilde** [comment]
Hey, is there a timeline to get this merged? I am using linkfield in a project which has a few of these and the experience is pretty clunky, I am a little worried the client might request we go back to linkable because of it.

**Comment by maxime-rainville** [comment]
I've got innovation week in a fortnight. I'll aim to get this across the line then.

FYI I've rebased the thing to work with CMS5

**Comment by maxime-rainville** [comment]
This PR is probably redundant now. 

The [AnyField](https://github.com/maxime-rainville/silverstripe-anyfield) module does everything this thing does in a more elegant fashion.

#76 add a few extra hooks to better integrate with AnyField.

**Comment by GuySartorelli** [comment]
Rebased and resolved merge conflict as part of additional work that's happening separately.

**Comment by GuySartorelli** [comment]
Closing in favor of https://github.com/silverstripe/silverstripe-linkfield/pull/120

**Pull request closed without merging**